flyio/caveats.md

```diff
@@ -0,0 +1,405 @@
+
+# Macaroons and Caveats at Fly
+
+This document covers the general mechanisms of Macaroons as they are used at
+Fly to do access control. It does not cover the cryptographic constructs.
+
+Macaroons are tokens. They are define in https://github.com/superfly/macaroon.
+At fly we pass them to services in `Authorization: Bearer <token>` HTTP headers.
+Several tokens can be passed in a single header, separated by commas. Access
+control is allowed if any one of the bearer tokens allows access.
+
+A Macaroon is extensible. Users that have a token can construct new tokens
+from them by adding Caveats. A Caveat is a restriction on the token's use.
+A token with no Caveats is all powerful, and tokens with Caveats wittle away
+at that power. Fly always issues tokens with an Org caveat which limits the
+token to operate on a single organization. A user in many organizations may
+pass in many tokens to operate across many organizations.
+
+Macaroons support "third party" caveats. These are special in a few ways.
+Third party caveats are "discharged" by an accompanying third party discharge
+token. A third party caveat in a token fails if there is no accompanying third party
+discharge passed along with the token. Fly only honors third party discharge tokens
+issued by certain trusted third party issuers. These are:
```

Comment:
We allow any third parties. We just don't trust attestation caveats from other TPs.

Reply:
if I remember correctly it can only verify third party discharges when there's a pre-established relationship, with a shared secret, so we can only verify and honor those third party caveats.

Reply:
The only difference is whether we ignore attestations:

https://github.com/superfly/macaroon/blob/d640a69989e6196f2c663b09cfaf4a8d11c6e236/macaroon.go#L375-L377

Reply:
changed the language slightly, but I didn't go into the details of what untrusted third party caveats mean.